From: Juho Vähä-Herttua <juhovh@iki.fi>
Date: Thu, 12 Jun 2008 17:41:49 +0000 (+0800)
Subject: OTHER: Move ALAC hack from avcodec to mp4 to make avcodec more general.
X-Git-Url: http://git.xmms.se/?p=xmms2-devel.git;a=commitdiff_plain;h=fae1d0cfd643e999d419162979b9c90d12a30002

OTHER: Move ALAC hack from avcodec to mp4 to make avcodec more general.
---

diff --git a/src/plugins/avcodec/avcodec.c b/src/plugins/avcodec/avcodec.c
index 37eaf73..4c4c794 100644
--- a/src/plugins/avcodec/avcodec.c
+++ b/src/plugins/avcodec/avcodec.c
@@ -40,6 +40,7 @@ typedef struct {
 	xmms_sample_format_t sampleformat;
 
 	gint bitrate;
+	gint samplebits;
 	const gchar *codec_id;
 	gpointer extradata;
 	gssize extradata_size;
@@ -148,6 +149,11 @@ xmms_avcodec_init (xmms_xform_t *xform)
 	                            "bitrate",
 	                            &data->bitrate);
 
+	/* ALAC and MAC require bits per sample field to be 16 */
+	xmms_xform_auxdata_get_int (xform,
+	                            "samplebits",
+	                            &data->samplebits);
+
 	ret = xmms_xform_auxdata_get_bin (xform,
 	                                  "decoder_config",
 	                                  &data->extradata,
@@ -162,12 +168,10 @@ xmms_avcodec_init (xmms_xform_t *xform)
 	data->codecctx->sample_rate = data->samplerate;
 	data->codecctx->channels = data->channels;
 	data->codecctx->bit_rate = data->bitrate;
+	data->codecctx->bits_per_sample = data->samplebits;
 	data->codecctx->extradata = data->extradata;
 	data->codecctx->extradata_size = data->extradata_size;
 
-	/* FIXME: this is for ALAC but can be a different value */
-	data->codecctx->bits_per_sample = 16;
-
 	if (avcodec_open (data->codecctx, codec) < 0) {
 		XMMS_DBG ("Opening decoder '%s' failed", codec->name);
 		goto err;
diff --git a/src/plugins/mp4/mp4.c b/src/plugins/mp4/mp4.c
index be34671..071bfb9 100644
--- a/src/plugins/mp4/mp4.c
+++ b/src/plugins/mp4/mp4.c
@@ -186,6 +186,9 @@ xmms_mp4_init (xmms_xform_t *xform)
 	xmms_xform_auxdata_set_bin (xform, "decoder_config", tmpbuf, tmpbuflen);
 	g_free (tmpbuf);
 
+	/* This is only for ALAC to set 16-bit samples, ignored for AAC */
+	xmms_xform_auxdata_set_int (xform, "samplebits", 16);
+
 	xmms_mp4_get_mediainfo (xform);
 
 	XMMS_DBG ("MP4 demuxer inited successfully!");
